core/src/main/scala/org/bykn/bosatsu/Evaluation.scala

```diff
@@ -415,39 +458,96 @@ case class Evaluation[T](pm: PackageMap.Typed[T], externals: Externals) {
               loop(as, itemFns, acc)
             }
 
-            if (dt.isStruct) {
-              // this is a struct, which means we expect it
-              { (arg: Value, acc: Env) =>
-                arg match {
-                  case p: ProductValue =>
-                    // this is the pattern
-                    // note passing in a List here we could have union patterns
-                    // if all the pattern bindings were known to be of the same type
-                    processArgs(p.toList, acc)
-
-                  case other =>
-                    // $COVERAGE-OFF$this should be unreachable
-                    val itemStr = items.mkString("(", ", ", ")")
-                    sys.error(s"ill typed in match (${ctor.asString}$itemStr\n\n$other\n\nenv: $acc")
-                    // $COVERAGE-ON$
+            if (dt.isNewType) {
+              // we erase this entirely, and just match the underlying item
+              itemFns.head
+            }
+            else if (dt.isStruct) {
+              if (itemsWild) noop
```

Comment:
possible further optimization is `itemsWild` could be a list of newly bound identifiers and you could check if they're actually referenced

Reply:
I’d rather do this optimization in the compiler when we normalize expressions.

I can see an argument for this being a compiler error actually.